### ISSUE TITLE
Use of "contains" at the top level

### ISSUE THREAD

**Issue by marcverhagen** [clamsproject/mmif#10]
This was proposed to contain a list (or probably a set) of annotation types in the MMIF views. You do not know what view those types are in, but it could be used for quick sniffing.

Some disadvantages:
- it makes MMIF a little bit more complex
- it introduces redundancies

And how much quicker is the sniffing? Without the "contains" on the top level all you would need to do something is loop over the views and collect the keys in the "contains" dictionary.

If we keep it at the top level I would like it better if it were placed inside the metadata dictionary.

**Comment by keighrim** [comment]
[Current implementation](https://github.com/clamsproject/clams-python-sdk/blob/18cf229203860c379dbb4f64d0849c469edc8ef8/clams/serialize/__init__.py#L44) of the top-level `contains` is actually a `dict` and stores the view ids of the attype. So in addition to *sniffing*, it also provides a quick way to find a view or views that contains a given annotation type. Putting it inside the metadata is a good idea, actually.

**Comment by marcverhagen** [comment]
Dictionary is better than a list or set. Still not convinced on how much easier it makes things though. 

Can the CLAMS platform be set up in such a way that the `metadata>contains` is created automatically from the views each time you create a MMIF structure? I do not see why not and it would deal with the redundancy if those data are generated automatically for convenience.

**Comment by keighrim** [comment]
We can provide python-sdk (or sdk for other languages) supporting that automatic insertion. However, if a developer does not use the *official* sdk then it's up to their implementation. I think we can also provide json schema and validator as part of sdk or independent tools to semi-force the *correct* MMIF syntax for developers. 

Top-level `contains` should not be a big burden in terms of disc space, even considering the possible redundancy. I think having it for sniffing and quick view resolution provides more advantages than harms.

**Comment by marcverhagen** [comment]
Oh, I don't think it is a burden in disk space at all, it's just that you typically do not want to express the same information twice.

But as long as we can view the "contains" in the top-level metadata as something generated from the views then I am okay with it.

**Comment by keighrim** [comment]
I would like to suggest a new name for this top-level `contains` just to avoid confusion with the `contains` in the `view` object. 
I'm thinking of `summary` and maybe we can add media to this dictionary, something link this; 

```json 
{
  "@context": "some-context", 
  "metadata": {
    "summary" : {
      "https://.../Video" : ["m1"], 
      "https://.../Text" : ["m2", "m3"], 
      "https://.../VideoSegment" : ["v1"], 
      "https://.../ASR" : ["v2", "m2"], # not so clear how to deal with this alignment yet
      "https://.../Token" : ["v3"], 
      "https://.../NamedEntity" : ["v4", "v5"]
    }
  },
  "media": [ {m1}, {m2}, {m3} ], 
  "views": [ {v1}, {v2}, {v3}, {v4}, {v5} ]
}
```

(In above I'm implicitly proposing "media types" defines as a part of MMIF namespace, preferably as vocab items. Which I'll give more thought before start a new issue maybe)

**Comment by marcverhagen** [comment]
I like using `summary` instead of `contains`. Still think that the SDK could just get all that stuff programmatically from the views though.

> (In above I'm implicitly proposing "media types" defines as a part of MMIF namespace, preferably as vocab items. Which I'll give more thought before start a new issue maybe)

Not sure what you mean by this but I'll await the new issue.

**Comment by angus-lherrou** [comment]
> Still think that the SDK could just get all that stuff programmatically from the views though.

I wonder how we would want to treat an incoming MMIF file with a `summary` that does not match the actual contents of the views. Would we want to reject it outright, or ignore the `summary` altogether and generate it correctly upon serialization? This potentially violates symmetry between `deserialize` and `serialize` but that might be okay if we just treat interaction with bad MMIF as undefined behavior.

**Comment by angus-lherrou** [comment]
Or are you proposing removing `summary`/top-level `contains` from the JSON schema altogether and just providing an interface for getting that information through the SDK?

**Comment by keighrim** [comment]
Earlier in the prototype, we designed the SDK to automatically *summarize* views whenever a new is added to the MMIF by `addView` method. That summarization was relying on a single helper function, and obviously a python developer has many different detours to nullify it. So that leaves us the fact that there's no particular way to validate or force the accuracy of the `summary` field, and with comments above on top of that fact I'm now turning more skeptical of the necessity of `summary` (format `contains`) field at all... Having it of course may save some I/O time in exchange to a trivial cost in disk storage, but with possible disparity between summary and actual content that we cannot control very well, I think it might leave a huge technical issues to the users of MMIF and the SDK.

**Comment by marcverhagen** [comment]
I would not at all be sad if we get rid of the top-level contains/summary. The redundancy and change of inconsistencies never sat well with me.

**Comment by keighrim** [comment]
closed via #36.
